vtex/faststore
# feat: Allow cookies management between FastStore API and frontend

## What's the purpose of this pull request?

Support cookies on Faststore API

## How it works?

We're using the request host to change the domain of the cookies, more details on the references

![Screenshot 2023-11-29 at 10 50 46](https://github.com/vtex/faststore/assets/19983991/4a4bbe3c-a893-4397-bf57-63b8ece785f5)
Cookies should be sent/received, as can be seen through this diagram - the numbers do not indicate an order, they are just to help with the following list:
1. [`Faststore API` **receives** cookies from `VTEX API` through the response.](https://github.com/vtex/faststore/blob/cd7de996b8c4fd0d6104678d53be18fdbcb55fed/packages/api/src/platforms/vtex/clients/fetch.ts#L21)
2. [`Faststore API` **sends** cookies to `VTEX API` through fetch.](https://github.com/vtex/faststore/blob/cd7de996b8c4fd0d6104678d53be18fdbcb55fed/packages/api/src/platforms/vtex/clients/fetch.ts#L14)
3. [`Faststore API` **sends** cookies to `storefront` adding it to the context storage.](https://github.com/vtex/faststore/blob/cd7de996b8c4fd0d6104678d53be18fdbcb55fed/packages/api/src/platforms/vtex/utils/cookies.ts#L4)
4. [`Faststore API` **receives** cookies from `storefront` that are sent through the request.](https://github.com/vtex/faststore/blob/4bac95f3d95edfab5695788d7db638f2cfbe4345/packages/core/src/pages/api/graphql.ts#L30)

## How to test it?

Inspect `validateCart` mutations on the network tab and see the `set-cookie` headers on the response. It is expected to have duplicated `set-cookie` values for the same cookie, only with different expiry times. We're letting the browser handle it, as only one of them will be actually set.

### Starters Deploy Preview

https://github.com/vtex-sites/starter.store/pull/300

## References

[Jira task](https://vtex-dev.atlassian.net/jira/software/projects/FS/boards/632?selectedIssue=FS-1463)
https://stackoverflow.com/questions/13851946/header-origin-vs-host/13871912#13871912
https://developer.mozilla.org/pt-BR/docs/Web/HTTP/Headers/Host
